platform-api/che-core-api-core/src/main/java/org/eclipse/che/api/core/rest/HttpRequestFactory.java

```diff
@@ -0,0 +1,65 @@
+/*******************************************************************************
+ * Copyright (c) 2012-2016 Codenvy, S.A.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Codenvy, S.A. - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.che.api.core.rest;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.List;
+import java.util.Map;
+
+import org.eclipse.che.commons.lang.Pair;
+
+import com.google.common.annotations.Beta;
+import com.google.inject.ImplementedBy;
+
+/**
+ * Issues general HTTP requests. For JSON requests use {@link HttpJsonRequestFactory}.
+ * 
+ * @author Tareq Sharafy
+ *
+ */
+@Beta
+@ImplementedBy(DefaultHttpJsonRequestFactory.class)
+public interface HttpRequestFactory {
+
+    /**
+     * Writes to the body stream of an HTTP request issued by
+     * {@link HttpRequestFactory#request(int, String, String, BodyWriter, Map, List)}.
+     * 
+     * @author Tareq Sharafy
+     *
+     */
+    public interface BodyWriter {
+        void writeTo(OutputStream out) throws IOException;
+    }
+
+    /**
+     * Issue a general HTTP request and return the response.
+     * 
+     * @param timeout
+     *            The read timeout. A non-positive value indicates using the default.
+     * @param url
+     *            The HTTP URL to connect to.
+     * @param method
+     *            The method. Null defaults to GET.
+     * @param body
+     *            A {@link BodyWriter} to use for writing the body. (optional)
+     * @param headers
+     *            A map of the header values. (optional)
+     * @param queryParams
+     *            A list of the query parameters to append to the URL. (optional).
+     * @return
+     * @throws IOException
+     */
+    HttpResponse request(int timeout, String url, String method, BodyWriter body, Map<String, String> headers,
```

Comment:
Is it possible to use builder style API of http client with body writer/reader that will implement streaming approach?